File: p11.py
# Day 11 - AOC 2020

def flatten(a):
  return "".join("".join(x) for x in a)

def flip1(m, x, y):
  count = 0
  d = [-1, 0, 1]
  for dx in d:
    for dy in d:
      if dx == 0 and dy == 0:
        continue
      nx = x + dx
      ny = y + dy
      if nx >= 0 and ny >= 0 and nx < len(m) and ny < len(m[0]):
        count += m[nx][ny] == '#'

  if count == 0 and m[x][y] == 'L':
    return True

  if count >= 4 and m[x][y] == '#':
    return True

  return False

def flip2(m, x, y):
  count = 0
  d = [-1, 0, 1]
  for dx in d:
    for dy in d:
      if dx == 0 and dy == 0:
        continue
      nx = x
      ny = y
      while True:
        nx += dx
        ny += dy
        if nx >= 0 and ny >= 0 and nx < len(m) and ny < len(m[0]):
          if m[nx][ny] == '.':
            continue
          count += m[nx][ny] == '#'
          break
        break

  if count == 0 and m[x][y] == 'L':
    return True

  if count >= 5 and m[x][y] == '#':
    return True

  return False

def sim(m):
  m_next = [['.' for _ in r] for r in m]
  for i in range(len(m)):
    for j in range(len(m[0])):
      if m[i][j] == '.':
        continue
      
      if flip2(m, i, j):
        m_next[i][j] = 'L' if m[i][j] == '#' else '#'
      else:
        m_next[i][j] = '#' if m[i][j] == '#' else 'L'

  stab = True

  if flatten(m) != flatten(m_next):
    stab = False

  return stab, m_next

def main():
    #file = open("sample_input.txt", "r")    
    file = open("input.txt", "r")
    m = [list(line.strip()) for line in file.readlines() if line.strip()]
    while True:
      stab, m = sim(m)
      if stab :
        break

    ans = flatten(m).count('#')

    print(ans)

    
if __name__ == "__main__":
    main()

File: p11.cpp

#include <iostream>
#include <map>
#include <queue>
#include <set>
#include <sstream>
#include <string>
#include <vector>
#include <fstream>
#include <algorithm>
#include <cstdlib>
#include <cmath>

#define I int64_t
#define UI std::size_t
#define FOR(i, b) for(UI (i) = 0; i < (b); i++)
#define P(s, c) std::cout << s << c << "\n"

std::vector<std::string> m;

bool flip2(I x, I y)
{
  I a[3] = {-1, 0, 1};
  I count = 0;
  FOR(i, 3)
  {
    FOR(j, 3)
    {
      if(a[i] == 0 && a[j] == 0)
      {
        continue;
      }

      auto nx = x;
      auto ny = y;

      // keep scanning in this direction
      while(true)
      {
        nx += a[i];
        ny += a[j];

        if(nx >= 0 && ny >= 0 && nx < m.size() && ny < m[0].size())
        {
          if(m[nx][ny] != '.')
          {
            count += m[nx][ny] == '#';
            break;
          }
          continue;
        }
        break;
      }      
    }
  }

  if(count == 0 && m[x][y] == 'L')
  {
    return true;
  }

  if(count >= 5 && m[x][y] == '#')
  {
    return true;
  }

  return false;
}

bool flip1(I x, I y)
{
  I a[3] = {-1, 0, 1};
  I count = 0;
  FOR(i, 3)
  {
    FOR(j, 3)
    {

      if(a[i] == 0 && a[j] == 0)
      {
        continue;
      }

      auto nx = x + a[i];
      auto ny = y + a[j];

      if(nx >= 0 && ny >= 0 && nx < m.size() && ny < m[0].size())
      {
        count += m[nx][ny] == '#';
      }
    }
  }

  if(count == 0 && m[x][y] == 'L')
  {
    return true;
  }

  if(count >= 4 && m[x][y] == '#')
  {
    return true;
  }

  return false;

}

bool sim()
{
  std::vector<std::string> m_next = m;
  FOR(i, m.size())
  {
    FOR(j, m[i].size())
    {
      if(m[i][j] == '.')
      {
        continue;
      }

      // use flip1 for part 1
      if (flip2(i, j))
      {
        m_next[i][j] = m[i][j] == '#' ? 'L' : '#';
      }
    }
  }

  bool stab = true;
  FOR(i, m.size())
  {
    if(m[i] != m_next[i])
    {
      stab = false;
    }
  }

  m = m_next;
  return stab;
}

int main() {
  
  std::ifstream file;
  //ile.open("sample_input.txt");
  file.open("input.txt");
  
  std::string line;
  while(std::getline(file, line))
  {
    m.emplace_back(line);
  }

  file.close();

  while(true)
  {
    auto stab = sim();
    if(stab)
    {
      break;
    }
  }

  I ans = 0;
  FOR(i, m.size())
  {
    FOR(j, m[i].size())
    {
      ans += m[i][j] == '#';
    }
  }

  P("Num occupied seats : ", ans);

  return 0;
}